src/resolve/resolved-dns-server.c

```diff
@@ -576,6 +576,29 @@ void dns_server_warn_downgrade(DnsServer *server) {
         server->warned_downgrade = true;
 }
 
+bool dns_server_limited_domains(DnsServer *server)
+{
+        DnsSearchDomain *domain;
+        bool domain_restricted = false;
+
+        /* Check if the server has route-only domains without ~., i. e. whether
+         * it should only be used for particular domains */
+        if (!server->link)
+                return false;
+
+        LIST_FOREACH(domains, domain, server->link->search_domains) {
+                if (domain->route_only) {
+                        domain_restricted = true;
+                        /* ~. means "any domain", thus it is a global server */
+                        if (streq(DNS_SEARCH_DOMAIN_NAME(domain), ".")) {
+                                return false;
+                        }
+                }
+        }
+
+        return domain_restricted;
```

Comment:
> >     - if none of the specified suffixes are routing-only, add the implicit ~. as in 1.
> I don't think we should do that by default, see above.

Argh, sorry, still confused about search expansion vs. routing -- yes, this *does* make sense indeed. With only "search" domains this should be a global name server. Just the search expansion should not happen on other links.